File: LevelUp_1May20/lecture001_Recursion/bits.cpp
#include <iostream>
#include <vector>
using namespace std;

int offToOn(int num, int k)
{
    int mask = (1 << k);
    return (num | mask);
}

int onToOff_01(int num, int k)
{
    int mask = (~(1 << k)); //1's compliment of mask.
    return (num & mask);
}

int onToOff_02(int num, int k)
{
    int mask = (1 << k); //1's compliment of mask.
    if ((num & mask) == 0)
        return num;
    else
        return (num ^ mask);
}

int countAllSetBits_02(unsigned_int num)
{
    int count = 0;
    while (num != 0)
    {
        if ((num & 1) != 0)
            count++;
        num >>= 1;
    }
}

//leetcode: 338===========================================

vector<int> countBits(int num)
{
    vector<int> ans(num+1,0);
    for(int i=1;i<=num;i++) ans[i]=ans[(i & (i-1))] + 1;
    return ans;
}

int main()
{
    int num = 123, k = 3;
    cout << (onToOff_01(num, k) == onToOff_02(num, k)) << endl;
    return 0;
}

File: LevelUp_1May20/lecture001_Recursion/bits.java
public class bits{
    public static void main(String[] args){

    }

    public static int countAllSetBits(int num){
        int count=0;
        for(int i=0;i<32;i++){
            int mask = (1<<i);
            if((num&mask)!=0) count++;
        }
        return count;
    }

    public static void countAllSetBits_02(int num){
        int count=0;
        while(num!=0){
            if((num & 1)!=0) count++;
            num >>>= 1;
        }
    }

    
    public static void countAllSetBits_03(int num){
        int count=0;
        while(num!=0){
            count++;
            num &=(num-1);
        }
    }

    public static int UniqueNumberInArray(int[] arr){
        int res=0;
        for(int ele: arr) res^=ele;
        return res;

    }
    
    //Leetcode : 231============================================
    public static bool isPowerOfTwo(int n) {
        return n>0 && (n & (n - 1))==0;
    }
}